packages/core/contracts/oracle/implementation/StakerSnapshot.sol

```diff
@@ -0,0 +1,242 @@
+    function executeUnstake() public {
+        _updateTrackers(msg.sender);
+        updateSnapshot(msg.sender);
+        VoterStake storage voterStake = voterStakes[msg.sender];
+        require(voterStake.unstakeTime != 0 && getCurrentTime() >= voterStake.unstakeTime, "Unstake time not passed");
+        uint256 tokensToSend = voterStake.requestUnstake;
+        if (tokensToSend > voterStake.cumulativeStaked) tokensToSend = voterStake.cumulativeStaked;
+
+        if (tokensToSend > 0) {
+            voterStake.cumulativeStaked -= tokensToSend;
+            cumulativeStaked -= voterStake.requestUnstake;
+            voterStake.requestUnstake = 0;
+            voterStake.unstakeTime = 0;
+            votingToken.transfer(msg.sender, tokensToSend);
+        }
+    }
+
+    // Send accumulated rewards to the voter. If the voter has gained rewards from others slashing then this is included
+    // here. If the total slashing is larger than the outstanding rewards then this method does nothing.
+    function withdrawRewards() public {
+        _updateTrackers(msg.sender);
+        VoterStake storage voterStake = voterStakes[msg.sender];
+
+        if (voterStake.outstandingRewards > 0) {
+            require(votingToken.mint(msg.sender, voterStake.outstandingRewards), "Voting token issuance failed");
+            voterStake.outstandingRewards = 0;
+        }
+    }
+
+    function exit() public {
+        executeUnstake();
+        withdrawRewards();
+    }
+
+    function _updateTrackers(address voterAddress) internal {
+        _updateReward(voterAddress);
+    }
+
+    // Calculate the reward per token based on last time the reward was updated.
+    function _updateReward(address voterAddress) internal {
+        rewardPerTokenStored = rewardPerToken();
+        lastUpdateTime = getCurrentTime();
+        if (voterAddress != address(0)) {
+            VoterStake storage voterStake = voterStakes[voterAddress];
+            voterStake.outstandingRewards = outstandingRewards(voterAddress);
+            voterStake.rewardsPaidPerToken = rewardPerTokenStored;
+        }
+    }
+
+    function outstandingRewards(address voterAddress) public view returns (uint256) {
+        VoterStake storage voterStake = voterStakes[voterAddress];
+
+        return
+            ((voterStake.cumulativeStaked * (rewardPerToken() - voterStake.rewardsPaidPerToken)) / 1e18) +
+            voterStake.outstandingRewards;
+    }
+
+    function rewardPerToken() public view returns (uint256) {
+        if (cumulativeStaked == 0) return rewardPerTokenStored;
+        return rewardPerTokenStored + ((getCurrentTime() - lastUpdateTime) * emissionRate * 1e18) / cumulativeStaked;
+    }
+
+    // Owner methods
+    function setEmissionRate(uint256 _emissionRate) public onlyOwner {
+        emissionRate = _emissionRate;
+    }
+
+    function setUnstakeCoolDown(uint256 _unstakeCoolDown) public onlyOwner {
+        unstakeCoolDown = _unstakeCoolDown;
+    }
+
+    // Snapshot methods
```

Comment:
Not exactly sure how we're planning to use this snapshot functionality, but it might be overkill for what we need. For instance, I don't think we need to track voters' balances all the way back in history, right? We only need to know their balance _in the most recent vote_, no?

Ideally, though, we might be able to get away _without snapshotting at all_ (this is half baked). Since all balance updates happen via the contract, we might be able to read the current staking balance for the purposes of tallying votes or accumulating rewards before taking any balance-modifying action on the user.

Reply:
this is not correct. you _do_ need to know the snapshots at all historic block rounds. the main reason is that when executing slashing you need to know how much the voter had staked at the round we are considering in the slashing.

you can see the two places in the slashing logic where this is required in the follow on PR.:

- https://github.com/UMAprotocol/protocol/pull/3981/files#r903470022
- https://github.com/UMAprotocol/protocol/pull/3981/files#r903469951